### PR TITLE
Add Basic authentication support

### PR DESCRIPTION
**Pull request by Chamindu36**
## Purpose
Provide authentication for product micro-gateway with basic authentication which uses usernames and passwords for authentication.

## Goals
Providing feature of enabling basic authentication security schema to product micro-gateway along with OAuth2 token based authentication. User will be benefited with using only OAuth2 token based authentication alone, using basic authentication alone and use both schemas at same time. 

## Approach
The ballerina source of the current micro-gateway is generated only for supporting OAuth2.0 token based authentication. We will introduce a new setup command flag as “-b” and “--security” which can be configured at  'setup' command as “--security=”values_of_securitySchemas”” to make the micro-gateway perform  based on basic authentication and as well as work with OAuth 2.0. When this command is provided, the ballerina source will be generated to support following user stories based on what are the values user is provided.Then at “service.mustache”  file, it has to check whether this isOptional and isRequired boolean values at “BasicAuth.java” class which has to be created under “Config.java” . Based on that value the source generation will be changed as follows at “service.mustache” file. Introduced a new file called “BasicAuth.bal” which perform the extraction of credentials from the request, validation and value settings for analytics inside it. Include this filter by default to the filter chain and if “AUTH_HEADER” is “basic” then request will process through this “BasicAuth.bal”. By using this design, this feature can provide support for basic authentication and OAuth2 based authentication at the same time and to only support for basic authentication only.
For further reference :
https://docs.google.com/document/d/11gMFm9yHlnDz2VXRBFSBCHjoLH3-F5w3zXpOaZG0ygc/edit?usp=sharing

## User stories

User Story | Micro-gateway that is only secured with oauth2.
-- | --
Overview | As an user, I need to secure my API only with oauth2.
Persona | API creator, API Publisher
Prerequisites | None
Narrative | User setup micro-gateway.Then user build micro-gateway and no additional steps are needed.Or user can setup micro gateway with --security ”oauth2” command
Acceptance Criteria | Micro-gateway that is secured with oauth2 token should be allowed to access, only if token validation passesMicro-gateway secured with this Oauth2 cannot be validated usernames and passwords.

User Story | Micro-gateway that is only secured with basic authentication.
-- | --
Overview | As an API creator/API Publisher,I need to secure my API only with usernames and passwords.
Persona | API creator, API Publisher
Prerequisites | Passwords converted into their SHA1 hash values.
Narrative | User setup micro-gateway with following flag and value.--security ”basic” Then user build micro-gateway and no additional steps are needed.User has to input authorized usernames and passwords to micro-gw.conf file.
Acceptance Criteria | Micro-gateway that is secured with basic auth should be allowed to access with authorized usernames and passwords.Micro-gateway with basic authentication should not be allowed to access with a valid oauth2 token.

User Story | Micro-gateway which is secured with basic authentication and with oauth2 token on same time.
-- | --
Overview | User need to secure micro-gateway with both oauth2 token or with usernames and passwords.
Persona | API creator, API Publisher
Prerequisites | Passwords converted into their SHA1 hash values.
Narrative | User setup micro-gateway with following flag and value.           --security ”oauth2,basic” Then user build micro-gateway and no additional steps are needed.User has to input authorized usernames and passwords to micro-gw.conf file.
Acceptance Criteria | Micro-gateway that is secured with both should be allowed to access with authorized usernames and passwords.Micro-gateway that is secured with both schemas should be allowed to access, only if token validation passes

## Documentation
Doc has Impact

## Automation tests
 - Unit tests 
testApiInvokeFailWithInvalidFormat
testApiInvokeFailWithInvalidPassword
testApiInvokeWithoutPassword
testApiInvokePassWithBasicAuth
testApiInvokeFailWithJWT
testApiInvokeWithJWT
testApiInvokeWithBasicAuth
testApiInvokeFailWithBasicAuth

 - Integration tests
1. APIInvokeWithBasicAuthTestCase
2. APIInvokeWithOAuth2andBasicAuthTestCase
3. APIInvokeWithOAuthTestCase

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 1]
```diff
@@ -444,4 +453,30 @@ private void populateHosts(String host) {
             throw new CLIInternalException("Error occurred while setting up URL configurations.");
         }
     }
+
+    public void setSecuritySchemas(String schemas) {
```
Comment (npamudika) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Shall we add java docs here?

**Comment by npamudika** [comment]
Can we please correct the PR title? "Adds" should be corrected as "Add"

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal [hunk 2]
```diff
@@ -24,18 +24,14 @@ import ballerina/time;
...
 @Description {value:"Representation of the Authentication filter"}
 @Field {value:"filterRequest: request filter method which attempts to authenticated the request"}
 public type AuthnFilter object {
 
     public OAuthnAuthenticator oauthnHandler;// Handles the oauth2 authentication;
-    public http:AuthnHandlerChain authnHandlerChain;
-
-
-    public new (oauthnHandler, authnHandlerChain) {}
+    public boolean oauth2Enabled = false;
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal]: Shall we rename the variable name to "isOAuth2Enabled" as it is a boolean variable?

File: components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal [hunk 3]
```diff
@@ -153,6 +153,30 @@ public function getAuthProviders() returns http:AuthProvider[] {
...
 }
 
+public function getBasicAuthProvider() returns http:AuthProvider[] {
+    http:AuthProvider basicAuthProvider = {
+        id: AUTHN_SCHEME_BASIC,
+        scheme: AUTHN_SCHEME_BASIC,
+        authStoreProvider: AUTH_PROVIDER_CONFIG
+    };
+    return [basicAuthProvider];
+}
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal]: Lets add an extra line after the method

File: components/micro-gateway-core/src/main/ballerina/gateway/utils/basicAuth.bal [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import ballerina/runtime;
+import ballerina/system;
+import ballerina/time;
+import ballerina/io;
+import ballerina/reflect;
+import ballerina/crypto;
+
+
+// MutualSSL filter
+
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/utils/basicAuth.bal]: Shall we remove this extra line?

File: components/micro-gateway-core/src/main/ballerina/gateway/utils/basicAuth.bal [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+@Description { value: "Representation of the BasicAuth filter" }
+
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/utils/basicAuth.bal]: This too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    private String prodToken, sandToken, jwtTokenProd, jwtTokenSand, expiringJwtTokenProd;
+
+    @BeforeClass
+    public void start() throws Exception {
+        String label = "apimTestLabel";
+        String project = "apimTestProject";
+        //get mock APIM Instance
+        MockAPIPublisher pub = MockAPIPublisher.getInstance();
+        APIDTO api = new APIDTO();
+        api.setName("PizzaShackAPI");
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java]: Shall we use a name specific for the test case while creating the API?

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal [hunk 7]
```diff
@@ -185,6 +196,12 @@ public type AuthnFilter object {
                         return false;
                     }
                 }
+            }else if (providerId == AUTHN_SCHEME_BASIC)
```
Comment (dushaniw) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal]: minor formatting issue 
```suggestion
            } else if (providerId == AUTHN_SCHEME_BASIC)
```

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + token);
+        org.wso2.micro.gateway.tests.util.HttpResponse response = HttpClientRequest
+                .doGet(getServiceURLHttp("/pizzashack/1.0.0/menu"), headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokePassWithBasicAuth() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        try {
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
+    }
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeWithoutPassword() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: This too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeFailWithInvalidPassword() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: This too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeFailWithInvalidFormat() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Here as well

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + token);
+        org.wso2.micro.gateway.tests.util.HttpResponse response = HttpClientRequest
+                .doGet(getServiceURLHttp("/pizzashack/1.0.0/menu"), headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeWithBasicAuth() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java]: Lets remove this additional line

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java [hunk 13]
```diff
@@ -136,7 +138,36 @@ private void invoke(String token, int responseCode) throws Exception {
         Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
     }
 
-    @AfterClass public void stop() throws Exception {
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeFailWithBasicAuth() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java]: This too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java [hunk 14]
```diff
@@ -136,7 +138,36 @@ private void invoke(String token, int responseCode) throws Exception {
...
+        //test endpoint
+        invokeBasic(basicAuthToken, 401);
+
+        try {
+            Thread.sleep(2000);
+        } catch (InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java]: Here as well

**Pull request merged**